File: solutions/python/queen-attack/1/queen_attack.py
class Queen:
    def __init__(self, row, column):
        if row < 0:
            raise ValueError("row not positive")
        if row > 7:
            raise ValueError("row not on board")
        if column < 0:
            raise ValueError("column not positive")
        if column > 7:
            raise ValueError("column not on board")
        self.row = row
        self.column = column

    def can_attack(self, other):
        if self.row == other.row and self.column == other.column:
            raise ValueError(
                "Invalid queen position: both queens in the same square"
            )
        d_row = abs(self.row - other.row)
        d_column = abs(self.column - other.column)
        return d_row == 0 or d_column == 0 or d_row == d_column

File: solutions/lua/queen-attack/1/queen-attack.lua
return function(position)
  assert(0 <= position.column and position.column <= 7)
  assert(0 <= position.row and position.row <= 7)
  return {
    row = position.row,
    column = position.column,
    can_attack = function(other)
      return position.row == other.row or
        position.column == other.column or
        position.row - position.column == other.row - other.column or
        position.row + position.column == other.row + other.column
    end,
  }
end